File: 1122.relative-sort-array.java
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;
import java.util.stream.Collectors;

/*
 * @lc app=leetcode id=1122 lang=java
 *
 * [1122] Relative Sort Array
 */

// @lc code=start
class Solution {
    public int[] relativeSortArray(int[] arr1, int[] arr2) {
        int[] count = new int[1001];
        for (int i = 0; i < arr1.length; ++i) {
            ++count[arr1[i]];
        }
        int cur = 0;
        for (var n : arr2) {
            while (count[n] != 0) {
                arr1[cur++] = n;
                --count[n];
            }
        }
        for (var i=0;i<count.length;++i) {
            while (count[i] != 0) {
                arr1[cur++] = i;
                --count[i];
            }
        }
        return arr1;
    }
}
// @lc code=end

// class Solution {
// public int[] relativeSortArray(int[] arr1, int[] arr2) {
// Map<Integer, Integer> comp_order = new HashMap<>();
// for (int i = 0; i != arr2.length; ++i) {
// comp_order.put(arr2[i], i);
// }
// List<Integer> list =
// Arrays.stream(arr1).boxed().collect(Collectors.toList());;
// Collections.sort(list, new MyComp(comp_order));
// return list.stream().mapToInt(i->i).toArray();
// }
// }

// class MyComp implements Comparator<Integer> {
// Map<Integer, Integer> comp_order;

// MyComp(Map<Integer, Integer> _comp_order) {
// comp_order = _comp_order;
// }

// @Override
// public int compare(Integer l, Integer r) {
// if (comp_order.containsKey(l) && comp_order.containsKey(r)) {
// return comp_order.get(l) - comp_order.get(r);
// } else if (comp_order.containsKey(l)) {
// return -1;
// } else if (comp_order.containsKey(r)) {
// return 1;
// } else {
// return l - r;
// }
// }
// }

File: cpp/1122.relative-sort-array.cpp
/*
 * @lc app=leetcode id=1122 lang=cpp
 *
 * [1122] Relative Sort Array
 */
#include <vector>
#include <unordered_map>
using namespace std;
// @lc code=start
class Solution
{
public:
    vector<int> relativeSortArray(vector<int> &arr1, vector<int> &arr2)
    {
        unordered_map<int, int> comp_order;
        for (int i = 0, len = arr2.size(); i != len; ++i)
        {
            comp_order[arr2[i]] = i;
        }
        sort(arr1.begin(), arr1.end(), [&comp_order](int &l, int &r)
             {
                 if (comp_order.count(l) && comp_order.count(r))
                 {
                     return comp_order[l] < comp_order[r];
                 }
                 else if (comp_order.count(l))
                 {
                     return true;
                 }
                 else if (comp_order.count(r))
                 {
                     return false;
                 }
                 else
                 {
                     return l < r;
                 }
             });
        return arr1;
    }
};
// @lc code=end
